MM-51228 - Calls: UX redesign

cpoile opened the pull request:
#### Summary
- Pretty much a complete revamp of the UI. Apologies for the single commit, the changes are all either interconnected or so little that a single commit for each would be a waste of time and not really helpful for the reviewer.
- This first pass should be for @tanmay-des -- I'm adding @streamer45 and @enahum for visibility now, and I'll tag you again for a code review once the UX pass is done.
- @tanmay-des I had to make guesses for many edge cases not covered in the designs, and for that reason it would be much more useful if you could run this for yourself on a simulator. I'll add a label to build an apk and you can install that on your android or an android sim. Let me know and I'll help you set that up. I'm not sure if you can do the same for an iOS sim, I'll look into it.

For now here are screenshots for you to look over. I tried to get a variety of combinations, but there are a bunch I'm surely missing:

Denim:
<img width="319" alt="image" src="https://user-images.githubusercontent.com/1490756/231897227-2e37c9c6-3ed1-4e72-9788-5e8b7119e545.png">
<img width="320" alt="image" src="https://user-images.githubusercontent.com/1490756/231897237-a6ea16ab-d4e1-4d78-a007-611a13490dfc.png">
<img width="325" alt="image" src="https://user-images.githubusercontent.com/1490756/231897249-e877dbfa-9f89-4940-9112-bc023d27e94c.png">
<img width="322" alt="image" src="https://user-images.githubusercontent.com/1490756/231897260-0da6a9a7-d11d-4d24-8ac5-333bcafbc1d4.png">
<img width="323" alt="image" src="https://user-images.githubusercontent.com/1490756/231897271-e3cd5920-0a41-44c6-9633-ffd7880abc28.png">
<img width="660" alt="image" src="https://user-images.githubusercontent.com/1490756/231897293-69f6cc20-b2c6-479c-a2dc-cbf638742cde.png">
<img width="664" alt="image" src="https://user-images.githubusercontent.com/1490756/231897310-3a8175ae-df8f-4d5b-97b1-aecae681aec6.png">

Onyx:
<img width="323" alt="image" src="https://user-images.githubusercontent.com/1490756/231897334-644e57f5-b38d-49bc-abae-e424e599c3b6.png">

Quartz:
<img width="320" alt="image" src="https://user-images.githubusercontent.com/1490756/231897362-7c82c9f1-8e08-4716-9461-9afaeb34c89f.png">

Indigo:
<img width="324" alt="image" src="https://user-images.githubusercontent.com/1490756/231897395-3a718f52-b857-4329-b8ba-6026ea061ebc.png">

Sapphire:
<img width="322" alt="image" src="https://user-images.githubusercontent.com/1490756/231897415-1052da10-71df-4225-880c-094adfbcac3e.png">


#### Ticket Link
- https://mattermost.atlassian.net/browse/MM-51228

#### Checklist

- [ ] Added or updated unit tests (required for all new features)
- [x] Has UI changes
- [x] Includes text changes and localization file updates
- [x] Have tested against the 5 core themes to ensure consistency between them.

#### Device Information
- Android: 13, Pixel 6
- iOS: 15.7, iPhone 7 plus

#### Release Note

```release-note
Calls: UI redesign.
```

mattermost-build commented:
Building app in separate branch.

mattermost-build commented:
Successfully building:  
https://circleci.com/gh/mattermost/mattermost-mobile/28689  
https://circleci.com/gh/mattermost/mattermost-mobile/28688

mattermost-build commented:
Artifact links:  
https://output.circle-artifacts.com/output/job/6b32f2ed-c304-46b3-89f5-87d3a684574c/artifacts/0/tmp/artifacts/Mattermost_Beta.ipa  
https://output.circle-artifacts.com/output/job/4ea2e850-b3ba-4ea6-9850-2b42d9a891eb/artifacts/0/tmp/artifacts/Mattermost_Beta.apk

mattermost-build commented:
Building app in separate branch.

mattermost-build commented:
Successfully building:  
https://circleci.com/gh/mattermost/mattermost-mobile/28746  
https://circleci.com/gh/mattermost/mattermost-mobile/28745

mattermost-build commented:
Artifact links:  
https://output.circle-artifacts.com/output/job/b3e0832a-3451-48af-97cc-c35e58662eeb/artifacts/0/tmp/artifacts/Mattermost_Beta.apk  
https://output.circle-artifacts.com/output/job/09cb5e42-b441-451e-83dc-d80b6484783b/artifacts/0/tmp/artifacts/Mattermost_Beta.ipa

mattermost-build commented:
Building app in separate branch.

mattermost-build commented:
Successfully building:  
https://circleci.com/gh/mattermost/mattermost-mobile/28759  
https://circleci.com/gh/mattermost/mattermost-mobile/28758

mattermost-build commented:
Artifact links:  
https://output.circle-artifacts.com/output/job/cb2f6844-bc3a-4f07-ad3e-6704ac6c5773/artifacts/0/tmp/artifacts/Mattermost_Beta.apk  
https://output.circle-artifacts.com/output/job/5cc59c42-17bf-43a0-9b1a-0d57a686fc0e/artifacts/0/tmp/artifacts/Mattermost_Beta.ipa

mattermost-build commented:
Building app in separate branch.

mattermost-build commented:
Successfully building:  
https://circleci.com/gh/mattermost/mattermost-mobile/28759  
https://circleci.com/gh/mattermost/mattermost-mobile/28758

mattermost-build commented:
Artifact links:  
https://output.circle-artifacts.com/output/job/cb2f6844-bc3a-4f07-ad3e-6704ac6c5773/artifacts/0/tmp/artifacts/Mattermost_Beta.apk  
https://output.circle-artifacts.com/output/job/5cc59c42-17bf-43a0-9b1a-0d57a686fc0e/artifacts/0/tmp/artifacts/Mattermost_Beta.ipa

mattermost-build commented:
Building app in separate branch.

mattermost-build commented:
Successfully building:  
https://circleci.com/gh/mattermost/mattermost-mobile/28817  
https://circleci.com/gh/mattermost/mattermost-mobile/28816

mattermost-build commented:
Artifact links:  
https://output.circle-artifacts.com/output/job/8063c067-8978-47fd-86ab-66d7163db15c/artifacts/0/tmp/artifacts/Mattermost_Beta.apk  
https://output.circle-artifacts.com/output/job/4467693a-412b-4adc-b48b-360c637aac90/artifacts/0/tmp/artifacts/Mattermost_Beta.ipa

cpoile commented:
@tanmay-des Changes are done, take a look at the new build and let me know. Thanks

cpoile commented:
@tanmay-des Here are the changes you mentioned in our DM: 
<img width="199" alt="image" src="https://user-images.githubusercontent.com/1490756/234973444-84bf5e47-53fa-4361-a8b6-8fdb79fd8a95.png">

@streamer45 @enahum We can start code review now, whenever you get a chance. Thanks.

enahum commented on app/products/calls/screens/call_screen/call_screen.tsx:
```diff
@@ -71,9 +82,10 @@ export type Props = {
 }
 
-const getStyleSheet = makeStyleSheetFromTheme((theme: Theme) => ({
+const getStyleSheet = ((theme: CallsTheme) => StyleSheet.create({
```
in another component you kept `makeStyleSheetFromTheme` with a `CallsTheme` argument, any reason why not to do the same here?

cpoile replied:
I... Have no idea. :) I think I was experimenting with different ways of doing it, and forgot to clean up. Thanks for catching it.

enahum commented on app/products/calls/screens/call_screen/call_screen.tsx:
```diff
@@ -298,11 +320,18 @@ const CallScreen = ({
     const [showControlsInLandscape, setShowControlsInLandscape] = useState(false);
     const [showReactions, setShowReactions] = useState(false);
+    const callsTheme = useMemo(() => makeCallsTheme(theme), [theme]);
+    const style = getStyleSheet(callsTheme);
```
as `getStyleSheet` is not using `makeStyleSheetFromTheme` this is not being memoized creating a new style on every render, either use `makeStyleSheetFromTheme` or memoize this

enahum commented on app/products/calls/screens/call_screen/call_screen.tsx:
this component is a bit large, I would definitely recommend to split it into smaller components

cpoile replied:
Agreed, it's on our board as part of code refactoring epic: https://mattermost.atlassian.net/browse/MM-52458

enahum commented on app/products/calls/screens/call_screen/raised_hand_banner.tsx:
```diff
@@ -0,0 +1,89 @@
+    const intl = useIntl();
+    const theme = useTheme();
+    const callsTheme = useMemo(() => makeCallsTheme(theme), [theme]);
+    const style = getStyleSheet(callsTheme);
```
not memoized, same as the other comment.

streamer45 reviewed (approved):
This is brilliant work, it looks so much better now 💯 A couple of nits I am seeing:

Vertical alignment seems a bit off here:
![image](https://user-images.githubusercontent.com/1832946/235267017-e3777321-a5e0-4ef1-9b41-a87ea90dc9a0.png)

We stopped using the border for user avatars on the other side, should we do the same here?

![image](https://user-images.githubusercontent.com/1832946/235267218-1318fcd4-f3a3-4522-a9b9-81e7c42f7797.png)

cpoile commented:
@streamer45 Thanks for spotting the alignment issue on the join call phone icon -- fixed. 

![13074](https://user-images.githubusercontent.com/1490756/235506999-95dbcf8e-c8ba-4527-9bb0-14e9e06560c0.jpg)


I think the noBorder looks better, go with this? (I vote yes.) @tanmay-des?  
I also noticed the color was slightly different than the webapp (but still not the same as `onlineIndicator` color), I fixed it.

![25727](https://user-images.githubusercontent.com/1490756/235507057-4f630886-4391-4fa8-9112-b2736467e914.jpg)

tanmay-des commented:
Looks great @cpoile, can we center-align the description container and maybe a add a little padding to the left. Also let's reduce the corner to 4px so that it matches the rest of the components. [Design link](https://www.figma.com/file/G1xHnuLDxW4YPacSY4BW08/Calls%3A-UX-Nits?node-id=1191-141105&t=VNFvARQJihwHmbDE-11)

![image](https://user-images.githubusercontent.com/9270746/235510266-32a8a872-3a2d-47a0-944e-8fd8a625f717.png)

cpoile commented:
@tanmay-des Cool, done. And here's a long name wrapping example too. Look good?

![IMG_E07D9A18F3A8-1](https://user-images.githubusercontent.com/1490756/235511948-b244d772-af22-4118-a548-ffa70e9aa2e0.jpeg)

![IMG_C17D2F67C261-1](https://user-images.githubusercontent.com/1490756/235511966-d978ab6c-4030-4638-9980-58549014dd8a.jpeg)

cpoile commented:
@tanmay-des Ok, done (I also adjusted the text sizing to match the designs). Also, here's a long-name to show the wrapping behavior.

![IMG_2C2E73A50ADA-1](https://user-images.githubusercontent.com/1490756/235512998-df374f7f-9b24-4818-8a72-1534d96455ca.jpeg)

![IMG_197C917AEAFF-1](https://user-images.githubusercontent.com/1490756/235513022-0d9ddb0f-3299-4597-8c4f-7059b3d877b3.jpeg)

tanmay-des commented:
Yeah it's perfect! @cpoile 
Great work

tanmay-des commented:
The root post looks great @cpoile 
No, I don't think we need that margin in the call banner.

streamer45 commented:
> ![image](https://user-images.githubusercontent.com/9270746/235510266-32a8a872-3a2d-47a0-944e-8fd8a625f717.png)

@tanmay-des A couple of questions on this:

1. Why are we using an outline icon in the button there? It seems inconsistent with the rest, web included.
2. Should the "Join call" text be bold?

cpoile commented:
@streamer45 1. Here is the solid. 2. It is bold, but it's 12px maybe that's why? @tanmay-des Here are the updated closeups:
![IMG_4AF0BA841694-1](https://user-images.githubusercontent.com/1490756/235933347-16533e75-1a0f-49be-8ab8-d639f7df07d7.jpeg)
![IMG_E2E8FF6B51A7-1](https://user-images.githubusercontent.com/1490756/235933379-6ddcaeef-34a4-4e60-a059-065b38641761.jpeg)

streamer45 commented:
Thanks @cpoile, I raised this because it looked very thin on some of your previous screenshots but maybe it's just the rendering. Icon looks great now.

tanmay-des commented:
It looks great @cpoile 
@streamer45 You're right about the Icon, I was about to bring that up as it appeared in [design team Icon audit](https://mattermost.atlassian.net/wiki/spaces/DES/pages/2449178625/Icons+Audit), we'll be replacing all the outlined icon instances to filled ones. 

It currently appears on the `Start Call` button on the channel headers and on mobile's more option (bottom sheet)

cpoile commented:
@tanmay-des Done (and the one we used for disable/enable calls in the channel info)

cpoile commented:
@enahum Could you quickly look at https://github.com/mattermost/mattermost-mobile/pull/7283/commits/3d5a27b60ba567e3171e1a1c7337e0928af20010 -- there are some code changes (more significant than css tweaking) which deserve another review. Thanks.

enahum commented:
@cpoile do we want this as part of 2.4 ? the release branch as already been cut

cpoile commented:
@enahum It would be a nice to have, but it's not functionally crucial.

amyblais commented:
I would be more comfortable leaving this to v2.5, unless this was specifically scheduled for v2.4.

enahum reviewed (approved):
Approving but would like to see my question answered

cpoile commented on app/screens/thread/thread.tsx:
```diff
@@ -20,14 +20,15 @@ import EphemeralStore from '@store/ephemeral_store';
     componentId: AvailableScreens;
     isCRTEnabled: boolean;
-    isInACall: boolean;
+    callsChannelState: CallsChannelState;
```
> how is passing the entire state impacting re-renders?
>
> Could it be better to always mount a component that checks for this values and renders null in the case is needed?

Unfortunately we do need these three booleans in the thread and channel screens, to determine whether to show the calls components, and where the place the `MoreMessages` banner. As for always mounting a component, we were doing that in the past but you suggested (rightly, I think) that we should not mount it if it's not needed.

So I think we're left with the question: should the three booleans be 1) wrapped in an `Observable<object>`, a 2) Object of Observables (which could be three-dot spread into the props object), or 3) individual `Observable<boolean>`s.

1 and 2 let us reduce code duplication. I tried both and chose 1 in https://github.com/mattermost/mattermost-mobile/commit/3d5a27b60ba567e3171e1a1c7337e0928af20010 . 3 means we have to duplicate some code, but I think it's clearer what's happening, and there's a chance that, because the booleans are directly observed, that rxjs/react might be able to do some value comparisons and cut out a rerender (where as 1, because it's an object, might cause a rerender because the object is different, even though only one of the the booleans changed).
 
So now I'm using 3 in https://github.com/mattermost/mattermost-mobile/pull/7283/commits/8ec56196b28f5597c86b6346cf6aa4173a22797b -- although there's duplication, I think is the best we can do to limit rerenders due to calls code.  

Incidentally, I did some debugging and it looks like calls only causes rerenders when one of those booleans changes, which is as expected and necessary. But `PostList` is being called multiple times (like, dozens of times) in longer threads -- maybe we should look at that, because it's pretty slow to render on my old iPhone 7 plus.

streamer45 commented:
> I would be more comfortable leaving this to v2.5, unless this was specifically scheduled for v2.4.

The only concern I have with doing this would be that the plugin is set to ship the UX web/desktop changes in 7.11 so keeping this for 2.5 will create some inconsistency as mobile side will essentially fall behind. If this is acceptable then fine but I also don't think there's anything critical here so it should be a safe enough addition.

/cc @cpoile @enahum @amyblais

amyblais commented:
Elias is cutting the RC builds for v2.4 tomorrow. If we're able to merge this by end of day today, that would be good.

streamer45 commented:
> Elias is cutting the RC builds for v2.4 tomorrow. If we're able to merge this by end of day today, that would be good.

Thanks Amy, that shouldn't be a problem. /cc @cpoile

streamer45 reviewed (commented):
A couple of things I noticed, please let me know if these are issues or known behaviours.

#### With screen sharing on, opening the reactions quick picker and react progressively overflows over the participants, partially covering it.

![image](https://user-images.githubusercontent.com/1832946/236219538-fc76b1fa-8d69-40f6-9dbe-3d86720d824b.png)

![image](https://user-images.githubusercontent.com/1832946/236219603-b9f99b3a-56e9-47db-a2dc-cd395cf3db04.png)

![image](https://user-images.githubusercontent.com/1832946/236219677-ca3f354e-ddd2-48a1-ad89-291de34edea1.png)

#### Reaction counter is allowed to increase indefinitely

![image](https://user-images.githubusercontent.com/1832946/236220173-93209fa1-ca82-480c-b5bd-060c8f15e5a1.png)

I'd expect users to be able to react only once per emoji at least.

cpoile commented:
@streamer45 Yes, this is expected, but could be different if we want. It's a tradeoff I made with respect to the screen sharing size and the participant list -- we don't have enough room with how the UI is designed, so I'm trying to maintain normal situation  (without reactions) legibility of the participant names, then trading off between the screen sharing size and the participant names when reactions appear (which I'm assuming are temporary). 
We could make different choices if we want -- do we make the screen sharing smaller and smaller in order to keep the participant names + host badge (which takes up quite a bit of vertical space)?

As for unlimited number of reactions -- it's always been like that, and exists on the webapp side as well (we just don't show numbers, we stack them instead).

streamer45 commented:
> @streamer45 Yes, this is expected, but could be different if we want. It's a tradeoff I made with respect to the screen sharing size and the participant list -- we don't have enough room with how the UI is designed, so I'm trying to maintain normal situation (without reactions) legibility of the participant names, then trading off between the screen sharing size and the participant names when reactions appear (which I'm assuming are temporary). We could make different choices if we want -- do we make the screen sharing smaller and smaller in order to keep the participant names + host badge (which takes up quite a bit of vertical space)?

Assuming we allow the user to full screen the video at any time my opinion is that we should retain all the information on screen and make the video smaller as a consequence. That said, I'll defer to @tanmay-des and we can revisit later, not necessarily in this PR.

> As for unlimited number of reactions -- it's always been like that, and exists on the webapp side as well (we just don't show numbers, we stack them instead).

You are right, somehow the counter tricked me there because it's far less clear who is actually reacting.

streamer45 commented:
A couple more nits/questions:

![image](https://user-images.githubusercontent.com/1832946/236228929-638f9331-210a-4d3e-9d00-2108c8a63ee6.png)

- Should the raise hand icon be the filled version?
- Should text be bold there?

cpoile commented:
> * Should the raise hand icon be the filled version?
> * Should text be bold there?

* Yes, fixed. We don't have a filled `off` version, so it's still outline.
* Yes, and the text size was wrong (for the buttons too). Fixed, thank you.

streamer45 commented:
> We don't have a filled off version, so it's still outline.

@tanmay-des Something to fix on your side perhaps?

streamer45 reviewed (approved):
Thanks, again :)

tanmay-des commented:
@streamer45 Thanks for catching those nits!
And yes, having the `right-hand-off` as outline is by design. Filled version of that icon looks odd that's why we are using the outline version

mattermost-build commented:
Cherry pick is scheduled.
